src/components/createFormattedComponent.tsx

```diff
@@ -3,13 +3,6 @@ import {invariantIntlContext} from '../utils';
 import {IntlShape, FormatDateOptions, FormatNumberOptions} from '../types';
 import {Context} from './injectIntl';
 
-type FormatFunctionName = 'formatDate' | 'formatTime' | 'formatNumber';
-type FormatOptions<T> = T extends 'formatDate'
-  ? FormatDateOptions
-  : T extends 'formatTime'
-  ? FormatDateOptions
-  : FormatNumberOptions;
-
 const displayNames: {
```

Comment:
```
enum DisplayName {
   formatDate = 'FormattedDate';
   formatTime = 'FormattedTime';
   formatNumber = 'FormattedNumber';
}

DisplayName['formatDate']// the value is "FormattedDate"
```

Reply:
I would personally avoid using enums

Reply:
why?

Reply:
please use https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions

Reply:
![image](https://user-images.githubusercontent.com/3842474/62922581-53554f80-bdde-11e9-9a0e-3f0a3563192a.png)

Reply:
Thanks for the hint (I didn't know about this feature). However, this project is currently using TS 3.3.

I plan to open an issue to list the tooling and setup improvements (e.g. latest TS, using RTL for tests, etc).

Reply:
Got

Reply:
@longlho use enum definition, `DisplayName`, I think it is the correct practice.

Reply:
`const enum` would work here I believe?

Reply:
Hmm sorry, I'm not sure I understand what you need me to do here 🤔

Reply:
```typescript
const enum DisplayName {
   formatDate = 'FormattedDate',
   formatTime = 'FormattedTime',
   formatNumber = 'FormattedNumber',
}
```

in component:
```
Component.displayName = DisplayName[name];
```

Reply:
I'm getting this error: 

<img width="756" alt="image" src="https://user-images.githubusercontent.com/1110551/62939518-c60bfe00-bdd1-11e9-8c7d-e92c2cbe07f1.png">

Reply:
replace `const enum` to `enum`

Reply:
Ah thanks, I never used `enum` so my knowledge of it is a bit limited.

https://github.com/formatjs/react-intl/pull/1415/commits/41d319fa1fe8ff415810645078f834e779d083f8